### DISCUSSION TITLE
What does "You cannot remove repositories from a security manager team" mean?

### DISCUSSION THREAD

**Discussion by justinmchase** [github/safe-settings#429] [Q&A] [answered]
I am getting this error in my logs and the teams are not updated

> You cannot remove repositories from a security manager team

I have a suborg file and a repo specific file which I would expect to merge into three teams. Plus there is a security team added to all repos from the enterprise level. The payload for this error appears to only be referencing the two teams from the suborg file.

**Comment by svg153** [comment, marked as answer]
Apparently, it's a bug here: https://github.com/github/safe-settings/issues/315#issuecomment-1282533790

**Reply by justinmchase** [reply]
Thanks I am planning on adding the security manager team under:

```yml
teams:
  include: [] # my teams
  exclude:
    - name: security-manager-team-name
```

Can you think of any reasons why that wouldn't solve it until the more elegant solution is available?

**Reply by svg153** [reply]
This config is not possible for teams (There is no code for that in [teams.js](https://github.com/github/safe-settings/blob/main-enterprise/lib/plugins/teams.js)). The `include` and `exclude` is only available for labels after #408. So, I think it is better, directly, do the solutions proposed [here](https://github.com/github/safe-settings/issues/315#issuecomment-1282533790).

**Reply by justinmchase** [reply]
Well I think that fix in that comment sounds perfect, I'd love to have this just work without thinking about it.

But [the documentation] does say this:

> It is possible to provide an include or exclude settings to restrict the collaborators, teams, labels to a list of repos or exclude a set of repos for a collaborator.

Is that just wrong?

**Reply by svg153** [reply]
As far as I know, I do think it's wrong. Let's see if @decyjphr can confirm.

**Reply by decyjphr** [reply]
Yes, that config setting is not correct. You can use `include` and `exclude` within `teams` but that is to include or exclude a list of repos from this setting. So for `org`-level  or `suborg`-level settings we can include or exclude a set of repos from this setting. This setting does not take a list of teams to include or exclude.

The error you are seeing in the logs is because there are `security manager` teams configured for the org, and safe-settings is trying to remove them because the `teams:` settings do not contain them. It is not the normal pattern to include these teams in `teams:` settings and the correct expectation that `safe-settings` ignores these. But the current behavior is it tries to remove any teams that are added to the repos but are not in the `settings`. However, security managers cannot be removed from repos using the API. 

This error does not cause any issues but we should enhance the code so it handles security manager teams better and doesn't try to remove them. Let me see if I can add this code fix, but if anyone has the bandwidth, the API for security managers is [here](https://docs.github.com/en/enterprise-cloud@latest/rest/orgs/security-managers?apiVersion=2022-11-28)

**Reply by justinmchase** [reply]
Ok, thats ok I didn't want to necessarily have include and exclude there anyway, I just misread the docs there. I'd say those docs if not exactly innacurate are easy to misread.

That being said, what I'm really hoping to know is if they "merge" all of the applicable files.

So if I have `settings.yml`, `suborgs/example.yml` and `repos/example.yml` which all apply to a specific repo and each one has 1 team in it, will the repo end up with 3 teams?

So for example if I just add the security managers into the `settings.yml` file will it append it to the list of teams all reapos are trying to set and be a workaround?

**Reply by martinm82** [reply]
From what I have seen, the `repos/example.yml` will take precedence and will ignore the ones defined in org and suborg settings.
